contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,897 @@
+interface LibraryComponentProps {
+  // Existing incorrect or incomplete typings
+}
+
+// Your code
+declare module 'library/path' {
+  interface LibraryComponentProps {
+    // Add or modify the typings
+    additionalProp: string;
+  }
+}
+```
+
+## Enums and Union Types
+
+Use **union types** instead of enums. Opt for union types whenever possible. Union types provide flexibility and can handle varying data types and dynamic value sets.
+
+```ts
+// Bad
+enum Colors {
+  Red,
+  Green,
+  Blue
+}
+
+// Good
+type Colors = 'red' | 'green' | 'blue';
+```
+You can also define a union type by deriving the type from an array. It's a great choice when seeking an easily iterable and directly usable array in code while keeping one source of truth.
+
+```ts
+// Bad
+const COLORS = ['red', 'green', 'blue'] as const;
+type Colors = 'red' | 'green' | 'blue';
+
+// Good
+import { TupleToUnion } from 'type-fest';
+
+const COLORS = ['red', 'green', 'blue'] as const;
+type Colors = TupleToUnion<typeof COLORS>;
+```
+
+Defining union types using an object offers several benefits. This approach it's a great choice when **enum-like identifiers are needed**, while maintaining one source of truth and improving readability.
+
+```ts
+// Bad
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = 'red' | 'green' | 'blue';
+
+// Good
+import { ValueOf } from 'type-fest';
+
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = ValueOf<typeof COLORS>
+```
+
+If using enum is necessary, utilize string enum. Do not use numeric and heterogenous enums as they auto increment its numeric values in unexpected ways, are hard to understand when they have mixed value types and have reverse mappings. This can help prevent potential bugs at runtime.
+
+```ts
+// Bad
+enum Colors1 {
+  Red,   // when not assigning anything, the value will start with 0.
+  Green, // and will increment, 1.
+  Blue,  // 2.
+}
+
+enum Colors2 {
+  Red = 10,            // 10.
+  Green,               // will increment to 11.
+  Blue = 30,           // 30.
+  Yellow,              // same, will increment to 31.
+  Magenta = 'magenta', // "magenta".
+  Brown = 100,         // 100.
+  Black,               // 101.
+}
+
+// Bad - We also end up having reverse mappings when using numeric values in enums.
+Colors2.Green; // 11.
+Colors2[11];   // "Green".
+
+// Good
+enum Colors {
+  Red = 'red',
+  Green = 'green',
+  Blue = 'blue',
+}
+```
+
+## Usage of type `any` and `unknown`
+
+Usage of `any` is forbidden unless there is a very good reason for it. The `any` type bypasses the TypeScript type-checking system, thus making the code less safe and more prone to bugs. Any usage of `any` type must be followed with a comment explaining why you are using it.
+
+```ts
+// Bad
+function processAPIResponse(response: any) {
+    // your code
+}
+
+// Good
+// As we are just going to persist the response in Onyx and
```

Comment:
I think this is a bad example because we absolutely should be typing everything we put in Onyx imo

Reply:
I made the example more generic, let me know what do you think! It's hard to come with a good example where any is justifiable 😆

Reply:
@blazejkustra thank you for coming up with examples for `any` usage. I know coming with examples is very hard 👍 

I don't want to ruin your effort, but I was wondering do we even need to put down an example where `any` is justifiable?

If I remember correctly, the approach we specified in the design doc is that whenever an engineer feels a need to go against a lint rule (`any` usage by forbidden in our ESLint rule configuration), they need to post a comment in `#expensify-open-source` and check if the situation they encounter really warrants an exception.

so instead of prescribing exceptions in the guideline, we can just say "Don't use `any`. If you think you need to, post explanation in the slack channel`

this is much simper IMO.

generally, if we have a lint rule that errors when the rule is broken, we can just say "Don't do it" in the guideline.
`ex. Don't use Enum.`

Reply:
I agree @hayata-suenaga , it's ok for me removing the example about `any` usage as it has been exceptionally hard to come up, and not including an example kinda help discouraging its usage even more imo. cc @blazejkustra